src/ice/agent.rs

```diff
@@ -761,63 +761,82 @@ impl IceAgent {
+                // The existence of this username is checked in the STUN parser.
+                let (local, remote) = message.split_username().unwrap();
+
+                let local_creds = self.local_credentials();
+                if local != local_creds.ufrag {
                     trace!(
-                        "Message rejected, remote user mismatch: {} != {}",
-                        remote,
-                        remote_creds.ufrag
+                        "Message rejected, local user mismatch: {} != {}",
+                        local,
+                        local_creds.ufrag
                     );
                     return false;
                 }
-            }
-        }
 
-        if message.is_successful_binding_response() {
-            let belongs_to_a_candidate_pair = self
-                .candidate_pairs
-                .iter()
-                .any(|pair| pair.has_binding_attempt(message.trans_id()));
+                if let Some(remote_creds) = &self.remote_credentials {
+                    if remote != remote_creds.ufrag {
+                        trace!(
+                            "Message rejected, remote user mismatch: {} != {}",
+                            remote,
+                            remote_creds.ufrag
+                        );
+                        return false;
+                    }
+                }
 
-            if !belongs_to_a_candidate_pair {
-                trace!("Message rejected, transaction ID does not belong to any of our candidate pairs");
-                return false;
+                do_integrity_check(true)
             }
-        }
+            (StunMethod::Binding, StunClass::Success | StunClass::Failure) => {
+                let belongs_to_a_candidate_pair = self
+                    .candidate_pairs
+                    .iter()
+                    .any(|pair| pair.has_binding_attempt(message.trans_id()));
+
+                if !belongs_to_a_candidate_pair {
+                    trace!("Message rejected, transaction ID does not belong to any of our candidate pairs");
+                    return false;
+                }
 
-        let (_, password) = self.stun_credentials(!message.is_response());
-        if !message.check_integrity(&password) {
-            trace!("Message rejected, integrity check failed");
-            return false;
+                do_integrity_check(false)
+            }
+            (StunMethod::Binding, StunClass::Unknown) => {
+                // Without a known class, it's impossible to know how to validate the message
+                trace!("Message rejected, unknown STUN class");
+                false
+            }
+            (StunMethod::Unknown, _) => {
+                // Without a known method, it's impossible to know how to validate the message
+                trace!("Message rejected, unknown STUN method");
+                false
+            }
```

Comment:
These are behavior changes from my previous enumeration. Since processing is so method and class-dependent, not knowing either means we can't really know how to even verify that this message is valid, or for this instance.